test/basiccontrollermap.c

```diff
@@ -0,0 +1,134 @@
+    SDL_Joystick *joy;
+    const char *type;
+    char guid[64];
+    int running = 1;
+
+    SDL_Init(SDL_INIT_JOYSTICK);
+    joy = SDL_JoystickOpen(0);
+    if (joy == NULL) {
+        SDL_Log("Couldn't open joystick 0: %s\n", SDL_GetError());
+        SDL_Quit();
+        return 0;
+    }
+
+    SDL_assert(SDL_JoystickFromInstanceID(SDL_JoystickInstanceID(joy)) == joy);
+    SDL_JoystickGetGUIDString(SDL_JoystickGetGUID(joy), guid, sizeof (guid));
+    switch (SDL_JoystickGetType(joy)) {
+    case SDL_JOYSTICK_TYPE_GAMECONTROLLER:
+        type = "Game Controller";
+        break;
+    case SDL_JOYSTICK_TYPE_WHEEL:
+        type = "Wheel";
+        break;
+    case SDL_JOYSTICK_TYPE_ARCADE_STICK:
+        type = "Arcade Stick";
+        break;
+    case SDL_JOYSTICK_TYPE_FLIGHT_STICK:
+        type = "Flight Stick";
+        break;
+    case SDL_JOYSTICK_TYPE_DANCE_PAD:
+        type = "Dance Pad";
+        break;
+    case SDL_JOYSTICK_TYPE_GUITAR:
+        type = "Guitar";
+        break;
+    case SDL_JOYSTICK_TYPE_DRUM_KIT:
+        type = "Drum Kit";
+        break;
+    case SDL_JOYSTICK_TYPE_ARCADE_PAD:
+        type = "Arcade Pad";
+        break;
+    case SDL_JOYSTICK_TYPE_THROTTLE:
+        type = "Throttle";
+        break;
+    default:
+        type = "Unknown";
+        break;
+    }
+
+    SDL_Log("Joystick\n");
+    SDL_Log("          name: %s\n", SDL_JoystickName(joy));
+    SDL_Log("          type: %s\n", type);
+    SDL_Log("           LED: %s\n", SDL_JoystickHasLED(joy) ? "yes" : "no");
+    SDL_Log("        rumble: %s\n", SDL_JoystickHasRumble(joy) ? "yes" : "no");
+    SDL_Log("trigger rumble: %s\n", SDL_JoystickHasRumbleTriggers(joy) ? "yes" : "no");
+    SDL_Log("          axes: %d\n", SDL_JoystickNumAxes(joy));
+    SDL_Log("         balls: %d\n", SDL_JoystickNumBalls(joy));
+    SDL_Log("          hats: %d\n", SDL_JoystickNumHats(joy));
+    SDL_Log("       buttons: %d\n", SDL_JoystickNumButtons(joy));
+    SDL_Log("   instance id: %d\n", SDL_JoystickInstanceID(joy));
+    SDL_Log("          guid: %s\n", guid);
+    SDL_Log("       VID/PID: 0x%.4x/0x%.4x\n", SDL_JoystickGetVendor(joy), SDL_JoystickGetProduct(joy));
+    
+    
+    SDL_Delay(1000);
+    SDL_Log("Checking Rumble functionality for 2 seconds\n");
+    SDL_JoystickRumble(joy, 0xAAFF, 0xAAFF, 2000);
+
+    while (running) {
+        SDL_WaitEvent(&event);
+        switch(event.type) {
+            case SDL_QUIT:
+                break;
+            case SDL_KEYDOWN:
+                if (event.key.keysym.sym == SDLK_ESCAPE)
+                    running = 0;
+                break;
+            case SDL_JOYAXISMOTION:
+                SDL_Log("Axis: %i %i\n", event.jaxis.axis, event.jaxis.value);
+                break;
+            case SDL_JOYBUTTONDOWN:
+                SDL_Log("Button: %i\n", event.jbutton.button);
+                break;
+            case SDL_JOYDEVICEREMOVED:
+                SDL_Log("Joy removed: %i\n", event.jdevice.which);
```

Comment:
The event loop should also quit when, for the removed device, `event.jdevice.which == SDL_JoystickInstanceID(joy)`.
(`testjoystick.c` does this)

Reply:
Applied!